azure-storage-queue/src/main/scala/akka/stream/alpakka/azure/storagequeue/AzureQueueSourceStage.scala

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.azure.storagequeue
+
+import com.microsoft.azure.storage.queue.{CloudQueue, CloudQueueMessage}
+import akka.stream.stage.{GraphStage, GraphStageLogic, OutHandler}
+import akka.stream.{Attributes, Outlet, SourceShape}
+import scala.collection.mutable.Queue
+import scala.concurrent.Future
+import scala.util.{Failure, Success}
+
+case class AzureQueueSourceSettings(
+    waitTimeSeconds: Int,
+    maxBufferSize: Int,
+    maxBatchSize: Int
+) {
+  require(maxBatchSize <= maxBufferSize, "maxBatchSize must be lower or equal than maxBufferSize")
+}
+
+object AzureQueueSourceSettings {
+  val default = AzureQueueSourceSettings(20, 100, 10)
+}
+
+class AzureQueueSourceStage(cloudQueue: CloudQueue, settings: AzureQueueSourceSettings)
+    extends GraphStage[SourceShape[CloudQueueMessage]] {
+  val out: Outlet[CloudQueueMessage] = Outlet("AzureCloudQueue.out")
+  override val shape: SourceShape[CloudQueueMessage] = SourceShape(out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+    val buffer = new Queue[CloudQueueMessage]
+
+    private val maxConcurrency = settings.maxBufferSize / settings.maxBatchSize
+    private var maxCurrentConcurrency = maxConcurrency
+    private var currentRequests = 0
+
+    private val successCallback = getAsyncCallback(handleSuccess)
+    private val failureCallback = getAsyncCallback(handleFailure)
+
+    private def canReceiveNewMessages: Boolean = {
+      val currentFreeRequests = (settings.maxBufferSize - buffer.size) / settings.maxBatchSize
+      currentFreeRequests > currentRequests && maxCurrentConcurrency > currentRequests
+    }
+
+    def receiveMessages(): Unit =
+      if (canReceiveNewMessages) {
+        implicit val executionContext = materializer.executionContext
+        currentRequests += 1
+
+        Future {
+          import scala.collection.JavaConverters._
+          val res = cloudQueue.retrieveMessages(settings.maxBatchSize, settings.waitTimeSeconds, null, null)
```

Comment:
There is only ever one pull until a graph stage pushes, so that means that as this stage only fetches on pull it does not do more concurrent requests. To run concurrent requests you'd have to introduce a buffer and fetch batches from the queue, there's a few different ways to do this but you should make sure it really makes sense before introducing that complexity.